src/merkle/path.rs

```diff
@@ -24,13 +24,31 @@ impl MerklePath {
 
     /// Computes the merkle root for this opening.
     pub fn compute_root(&self, index_value: u64, node: Word) -> Word {
-        let mut index = NodeIndex::new(self.depth(), index_value);
-        self.nodes.iter().copied().fold(node, |node, sibling| {
-            // compute the node and move to the next iteration.
-            let input = index.build_node(node.into(), sibling.into());
-            index.move_up();
-            Rpo256::merge(&input).into()
-        })
+        self.compute_parents(index_value, node)
+            .last()
+            .unwrap_or(node)
+    }
+
+    /// Compute parent nodes based on the provided data
+    pub fn compute_parents(&self, index_value: u64, node: Word) -> impl Iterator<Item = Word> + '_ {
+        struct State {
+            index: NodeIndex,
+            curr_node: RpoDigest,
+        }
```

Comment:
Isn't overkill to add a struct here when this could easily be a tuple?

To give further context; I think it's always best to create wrapping structs when we have exposed tuples. But in this case, this is completely internal and the tuple won't ever leak

Reply:
The type annotations are required for the `into` calls.

Reply:
Yeah, looking over these two alternatives there is no advantage to either. Going to keep the original one since it is already there.